### PR TITLE
[SDL] Orchestrator updates the blob metadata with SHA512 before the copy. 

### PR DESCRIPTION
**Pull request by zhhyu**
Orchestrator updates the blob metadata with SHA512 before the copy.
Towards fixing MD5 task:
https://github.com/nuget/engineering/issues/2377
https://github.com/nuget/engineering/issues/2672

Verify it locally.

### CODE REVIEW

File: src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs [hunk 1]
```diff
@@ -286,6 +289,8 @@ await _packageFileService.DoesValidationSetPackageExistAsync(validationSet))
                     validationSet.PackageNormalizedVersion,
                     validationSet.ValidationTrackingId);
 
+                metaData = await _packageFileService.UpdatePackageBlobMetadataInValidationAsync(validationSet);
```
Comment (cristinamanum) [inline src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs]: This line and line 272 seems to be duplicate. The validationSet seems not changed on the two is/else branches . Can it be only one call?

Comment (zhhyu) [reply src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs]: Line 272 will copy the package from the validation-set folder inside the validation container to package container, and line 292 will copy the package from the validation container to package container.
Actually I am not very sure when the second case will happen, but based on the copy source and destination, we should always update the metadata on the source location.

File: src/NuGet.Services.Validation.Orchestrator/PackageStreamMetadataAndCopiedStatusWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using NuGetGallery.Packaging;
+
+namespace NuGet.Services.Validation.Orchestrator
+{
+    public class PackageStreamMetadataAndCopiedStatusWrapper
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/PackageStreamMetadataAndCopiedStatusWrapper.cs]: This type can be nested and private in `EntityStatusProcessor` right?

The name wrapper is a little confusing here. You're not wrapping logic here; it's just a simple container. Consider a name based on the method that is returning it, e.g. `UpdatePublicPackageResult`.

File: src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs]: This is a dead code path. It was used when we transitioned to having a copy of the package per validation set. See the log message.

File: src/NuGet.Services.Validation.Orchestrator/IValidationPackageFileService.cs [hunk 3]
```diff
@@ -114,15 +114,37 @@ public interface IValidationFileService
...
+        /// <summary>
+        /// Updates package blob metadata in the package container.
+        /// </summary>
+        /// <param name="validationSet">A validationSet that will identify the package that will have its blob metadata updated.</param>
+        /// <returns>A task that represents the asynchronous operation.
+        /// The task result (<see cref="Task{PackageStreamMetadata}.Result" />) returns
+        /// a <see name="PackageStreamMetadata" />.</returns>
+        /// <exception cref="Microsoft.WindowsAzure.Storage.StorageException">Thrown if the blob has changed between
+        /// successive read and write operations.</exception>
+        Task<PackageStreamMetadata> UpdatePackageBlobMetadataInPackageAsync(PackageValidationSet validationSet);
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/IValidationPackageFileService.cs]: Is this used anymore?

File: src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs]: Checked AI. The log on 291 has not happened in the past 90 days, as expected.

File: src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs [hunk 4]
```diff
@@ -268,6 +269,14 @@ await _packageFileService.DoesValidationSetPackageExistAsync(validationSet))
                         validationSet.PackageETag);
                 }
 
+                metaData = await _packageFileService.UpdatePackageBlobMetadataInValidationSetAsync(validationSet);
+
+                _logger.LogInformation(
+                        "Update the blob metadata of validation set {ValidationSetId} package {PackageId} {PackageVersion} with SHA512 hash value",
+                        validationSet.ValidationTrackingId,
+                        validationSet.PackageId,
+                        validationSet.PackageNormalizedVersion);
```
Comment (loic-sharma) [inline src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs]: Nit: this seems to be indented by an extra 4 spaces

Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs]: Same for the comment on line 300

---
In reply to: [321360882](https://github.com/NuGet/NuGet.Jobs/pull/809#discussion_r321360882) [](ancestors = 321360882)

File: src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs [hunk 5]
```diff
@@ -268,6 +269,14 @@ await _packageFileService.DoesValidationSetPackageExistAsync(validationSet))
                         validationSet.PackageETag);
                 }
 
+                metaData = await _packageFileService.UpdatePackageBlobMetadataInValidationSetAsync(validationSet);
+
+                _logger.LogInformation(
+                        "Update the blob metadata of validation set {ValidationSetId} package {PackageId} {PackageVersion} with SHA512 hash value",
```
Comment (loic-sharma) [inline src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs]: We may choose to add more metadata to blobs in the future. Consider making this log message a little more generic:

```suggestion
                        "Updated the blob metadata of validation set {ValidationSetId} package {PackageId} {PackageVersion}",
```

Same comment for line 301

File: src/NuGet.Services.Validation.Orchestrator/IValidationPackageFileService.cs (hunk 3 above)
Comment (zhhyu) [reply src/NuGet.Services.Validation.Orchestrator/IValidationPackageFileService.cs]: Delete it if no need now.

File: src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs (hunk 1 above)
Comment (zhhyu) [reply src/NuGet.Services.Validation.Orchestrator/EntityStatusProcessor.cs]: Will send another PR to remove the dead path.

**Pull request merged**